packages/@react-spectrum/dialog/src/DialogTrigger.tsx

```diff
@@ -144,13 +144,48 @@ export {_DialogTrigger as DialogTrigger};
 
 function PopoverTrigger({state, targetRef, trigger, content, hideArrow, isKeyboardDismissDisabled, ...props}) {
   let triggerRef = useRef<HTMLElement>();
-
   let overlayRef = useRef<DOMRefValue<HTMLDivElement>>();
+  let isCloserPadding = false;
+
+  // For popover triggers close to the edge, decrease the container padding allowing popovers to
+  // move closer to their trigger.
+  if (triggerRef?.current) {
```

Comment:
this is all during a render? that seems wrong, especially on first render. `useLayoutEffect` is for when you need to make adjustments based on dom layout, your ref will be defined at that time

calculations can be performed during render, based on state, but we shouldn't be reading from the dom in render. See `Arrow` or your work in `PopoverWrapper`

Reply:
Maybe we need to talk. When I moved this logic into a `useLayoutEffect` (and a `useEffect`) it only runs on page load. It needs to run on trigger which is what happens without the useLayoutEffect.